cpp/src/Glacier2Lib/SessionHelper.cpp

```diff
@@ -683,48 +682,21 @@ SessionHelperI::dispatchCallback(const Ice::DispatcherCallPtr& call, const Ice::
     }
 }
 
-namespace
-{
-
-class DispatcherCallWait : public Ice::DispatcherCall
-{
-
-public:
-
-    DispatcherCallWait(IceUtilInternal::CountDownLatch& cdl, const Ice::DispatcherCallPtr& call) :
-        _cdl(cdl),
-        _call(call)
-    {
-    }
-
-    virtual void
-    run()
-    {
-        _call->run();
-        _cdl.countDown();
-    }
-
-private:
-
-    IceUtilInternal::CountDownLatch& _cdl;
-    const Ice::DispatcherCallPtr _call;
-};
-
-}
-
 void
 SessionHelperI::dispatchCallbackAndWait(const Ice::DispatcherCallPtr& call, const Ice::ConnectionPtr& conn)
 {
     if(_initData.dispatcher)
     {
-        IceUtilInternal::CountDownLatch cdl(1);
-        auto callWait = make_shared<DispatcherCallWait>(cdl, call);
-        _initData.dispatcher([callWait]()
+        promise<void> dispatchPromise;
+
+        _initData.dispatcher(
+            [&dispatchPromise, call = std::move(call)]()
             {
-                callWait->run();
+                call->run();
+                dispatchPromise.set_value();
```

Comment:
Is this run call guaranteed to never throw?

Reply:
I don't think so. This keeps the previous behavior. Maybe is more correct to catch the exception and use it to complete the promise.

Reply:
As far as I can tell, Ice::DispatcherCallPtr is a C++98 API. Why are we still using it?

https://doc.zeroc.com/ice/3.7/client-server-features/the-ice-threading-model/dispatching-requests-to-user-threads#227755398f6b9995a15f4b74982e395cc4afef7b

Reply:
Seems SessionHelper is the only place where we still use it. I will remove it.

Reply:
According to the Ice Manual:
> The call given to the dispatcher takes no parameters, returns nothing, and never throws any exception back to the dispatcher.

We should mark is noexcept. So I wouldn't worry about the exception is throws.

Reply:
And in this particular case, all the "callbacks" are SessionHelper code, right?

Reply:
The callbacks call to the user provided SessionCallback.